uvm/impl/com/untangle/uvm/DeviceTableImpl.java

```diff
@@ -85,58 +89,149 @@ public int size()
+            throw new IllegalArgumentException("Invalid auto device removal threshold");
+
+        logger.info("Setting new device settings: {}", newSettings);
+        ConcurrentHashMap<String, DeviceTableEntry> oldDeviceTable = deviceTable;
+        deviceTable = new ConcurrentHashMap<>();
+
+        // For each entry, If it isn't in the table, create new entry. else keep as it is.
+        newSettings.getDevices()
+            .stream()
+            .filter(entry -> null != entry.getMacAddress())
+            .forEach(entry -> {
+                DeviceTableEntry existingEntry = oldDeviceTable.get(entry.getMacAddress());
+                if (existingEntry != null) {
+                    existingEntry.copy(entry);
+                    deviceTable.put(existingEntry.getMacAddress(), existingEntry);
+                } else {
+                    addDevice(entry);
+                    entry.enableLogging();
+                }
+            });
+        logger.info("Loaded devices from file. ({} entries)", deviceTable.size());
+
+        // If configs have been changed run devices removal
+        if (null != devicesSettings && newSettings.isAutoDeviceRemove()) {
+            if (!devicesSettings.isAutoDeviceRemove() || devicesSettings.getAutoRemovalThreshold() != newSettings.getAutoRemovalThreshold()) {
+                devicesSettings.setAutoDeviceRemove(true);
+                devicesSettings.setAutoRemovalThreshold(newSettings.getAutoRemovalThreshold());
+                autoDeleteDevices(true);        // skip save as we will be saving the settings later in this method.
+            }
+        }
+
+        // Change current settings to new settings
+        devicesSettings = newSettings;
+        // Save the settings. Skip last save time check
+        saveDevicesSettings(false);
     }
 
     /**
-     * Set the list of devices
-     *
-     * @param newDevices
-     *        The list of devices
+     * Saves the settings to devices.js file
+     * Checks if last save time is older than WAIT_BETWEEN_TWO_SAVES
+     * if lastSaveTimeCheck is true waits for remaining time else continues
+     * @param lastSaveTimeCheck if false, last save time check is skipped
      */
-    public synchronized void setDevices(LinkedList<DeviceTableEntry> newDevices)
-    {
-        ConcurrentHashMap<String, DeviceTableEntry> oldDeviceTable = this.deviceTable;
-        this.deviceTable = new ConcurrentHashMap<>();
+    public void saveDevicesSettings(boolean lastSaveTimeCheck) {
+        // If we just recently saved, within 60 seconds wait.
+        synchronized (lock) {
+            long currentTime = System.currentTimeMillis();
+            long timeSinceLastSave = currentTime - lastSaveTime;
+            long waitTime;
+            if (lastSaveTimeCheck && timeSinceLastSave < WAIT_BETWEEN_TWO_SAVES) {
+                logger.info("Saved recently, waiting...");
```

Comment:
We are also dynamically adding the devices by intercepting client traffic. So to avoid frequent save after each add device calls we need to wait.
Replaced lock logic with synchronized method